spacesuit/src/value.rs

```diff
@@ -126,10 +126,28 @@ impl Into<Scalar> for SignedInteger {
 }
 
 impl Add for SignedInteger {
-    type Output = SignedInteger;
+    type Output = Option<SignedInteger>;
 
-    fn add(self, rhs: SignedInteger) -> SignedInteger {
-        SignedInteger(self.0 + rhs.0)
+    fn add(self, rhs: SignedInteger) -> Option<SignedInteger> {
+        let s = self.0 + rhs.0;
+        if s <= (u64::max as i128) && s >= -(u64::max as i128) {
+            Some(SignedInteger(s))
+        } else {
+            None
+        }
+    }
+}
+
+impl Mul for SignedInteger {
+    type Output = Option<SignedInteger>;
+
+    fn mul(self, rhs: SignedInteger) -> Option<SignedInteger> {
+        let p = self.0 * rhs.0;
```

Comment:
Sorry, i probably was not clear enough in https://github.com/interstellar/slingshot/pull/106#discussion_r259173002

> However, `(2^64-1)^2` is larger than `2^127-1`, so we must use `checked_mul` here.

So we not only need to check that `p` is within `±u64::max`, but also that the multiplication did not overflow i128.

```suggestion
        if let Some(p) = self.0.checked_mul(&rhs.0) {
```

Reply:
```rust
if check i128 overflow {
    if check ±u64 overflow {
        Some(...)
    } else {
        None
    }
} else {
    None
}
```